Add DateTime content type

timkoopmans50 opened the pull request:
Added datetime content type and configuration.

| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | no
| Related issues/PRs | no
| License | MIT
| Documentation PR | https://github.com/sulu/sulu-docs/pull/533

#### What's in this PR?

Adds DateTime content type.

#### Why?

There was a Time and a Date content type but no datetime content type. 

#### Example Usage

in config/templates/pages/{yourpagename}.xml inside <properties> : 
```
        <property name="my_datetime" type="datetime">
            <meta>
                <title lang="en">DateTime</title>
                <title lang="de">DateTime Duits</title>
            </meta>
        </property>

```

#### To Do

fix failed tests?

nnatter commented on src/Sulu/Bundle/PageBundle/Content/Types/DateTime.php:
```diff
@@ -0,0 +1,74 @@
+        if (null != $value) {
+            $value = \DateTime::createFromFormat(self::format, $value);
+            $new = new \DateTime();
+            $format = $new->format(self::format);
```
not sure if this is used anywhere

nnatter commented on src/Sulu/Bundle/PageBundle/Content/Types/DateTime.php:
```diff
@@ -0,0 +1,74 @@
+        return $value;
+    }
+
+    public function getDefaultParams(PropertyInterface $property = null)
```
i suspect these params were used in 1.6, but are not used anymore in 2.0. so i think i would just remove them 🙂

nnatter commented:
Hey,
I rebased this PR and implemented the suggestions. Hope that is okay for you.
Thanks again for your effort!

timkoopmans50 commented:
> Hey,
> I rebased this PR and implemented the suggestions. Hope that is okay for you.
> Thanks again for your effort!

Sure! When is this due to be reviewed / merged?

nnatter commented:
I dont know, I am sorry. Hopefully before the next release 🙂 @sulu/core-team

alexander-schranz commented:
@timkoopmans50 @nnatter this will sadly not be in the next release as we are currently preparing it. We will make sure that we will have a look at it after the release.

danrot commented on src/Sulu/Bundle/PageBundle/Content/Types/DateTime.php:
```diff
@@ -0,0 +1,64 @@
+ */
+class DateTime extends SimpleContentType
+{
+    const format = 'Y-m-d\TH:i:s';
```
Constants should be uppercase:

```php
const FORMAT = '...';
```

danrot commented on src/Sulu/Bundle/PageBundle/Content/Types/DateTime.php:
```diff
@@ -0,0 +1,64 @@
+    ) {
+        $value = $property->getValue();
+        if (null != $value) {
+            $value = \DateTime::createFromFormat(self::format, $value);
```
Use `static` instead of `self`, which allows to override the value in a sub class.

timkoopmans50 commented:
> Once you've fixed these minor issues I am happy to test it. If everything works I'll merge it right away 🙂

Thank you, I've just pushed these changes to the feature branch.

danrot commented on src/Sulu/Bundle/PageBundle/Content/Types/DateTime.php:
```diff
@@ -0,0 +1,64 @@
+            $propertyValue = $node->getPropertyValue($property->getName());
+
+            if ($propertyValue instanceof \DateTime) {
+                $value = $propertyValue->format(static::FORMAT);
```
I think formatting the value like this might be a bit impractical for the website output... In the twig template you'll receive a string value like `"2020-08-21T23:05:00"`. This  makes it harder than it should be to align the output with the given language. I would suggest we override the `getContentData` method of the `SimpleContentType` class, and return a `DateTime` value there.

danrot replied:
Oh, and please don't forget to write a test for that as well :slightly_smiling_face:

nnatter replied:
done 🙂

danrot commented on config/templates/pages/example.xml:
```diff
@@ -196,6 +196,13 @@
             </meta>
         </property>
 
+        <property name="datetime" type="datetime">
```
Please also add an output for that in the `example.html.twig` file

nnatter replied:
added output for all missing properties

alexander-schranz commented on templates/pages/example.html.twig:
```diff
@@ -32,6 +32,7 @@
     <p>{{ content.color }}</p>
     <p>{{ content.time }}</p>
     <p>{{ content.date }}</p>
+    <p>{{ content.datetime }}</p>
```
As this was changed to datetime (not sure if a good idea). This twig file need to be change to the following:

```twig
{{ content.datetime  ? content.datetime.format('Y-m-d H:i') : '' }}
```

danrot replied:
How else would you allow to properly format the string otherwise? The content type itself cannot know if you want to write e.g. the month in a numeric or spelled out way. And this way also [twig functionality can be reused](https://twig.symfony.com/doc/2.x/filters/format_datetime.html).

nnatter replied:
adjusted 🙂

alexander-schranz commented on templates/pages/example.html.twig:
```diff
@@ -57,10 +58,14 @@
         <p>{{ content.single_account_selection.name }}</p>
     {% endif %}
 
+    <p>{{ content.account_selection | map(account => account.name) | join(', ') }}</p>
```
As we follow the symfony code style, there should be no spaces between the filters `|` /cc @nnatter

danrot commented on src/Sulu/Bundle/PageBundle/Content/Types/Date.php:
```diff
@@ -20,6 +20,8 @@
  */
 class Date extends SimpleContentType
 {
+    const FORMAT = 'Y-m-d\TH:i:s';
```
Why did that format change? This seems like the format for the `DateTime`, doesn't it?

nnatter replied:
you are right, thats my fault. fixed 🙂
